packages/next/src/server/app-render/collect-segment-data.tsx

```diff
@@ -0,0 +1,249 @@
+  const segmentBufferMap = new Map<string, Buffer>()
+  await collectSegmentDataImpl(
+    routeTree,
+    fullPageDataBuffer,
+    clientModules,
+    serverConsumerManifest,
+    [],
+    '',
+    segmentBufferMap
+  )
+
+  return segmentBufferMap
+}
+
+async function collectSegmentDataImpl(
+  route: FlightRouterState,
+  fullPageDataBuffer: Buffer,
+  clientModules: ManifestNode,
+  serverConsumerManifest: any,
+  segmentPath: Array<[string, Segment]>,
+  segmentPathStr: string,
+  segmentBufferMap: Map<string, Buffer>
+): Promise<void> {
+  const children = route[1]
+  for (const parallelRouteKey in children) {
+    const childRoute = children[parallelRouteKey]
+    const childSegment = childRoute[0]
+    const childSegmentPath = segmentPath.concat([
+      [parallelRouteKey, childSegment],
+    ])
+    const childSegmentPathStr =
+      segmentPathStr +
+      '/' +
+      encodeChildSegmentAsFilesystemSafePathname(parallelRouteKey, childSegment)
+    await collectSegmentDataImpl(
+      childRoute,
+      fullPageDataBuffer,
+      clientModules,
+      serverConsumerManifest,
+      childSegmentPath,
+      childSegmentPathStr,
+      segmentBufferMap
+    )
+  }
+
+  // Create a new Flight response that contains data only for this segment.
+  try {
+    const segmentStream = renderToReadableStream(
+      <PickSegment
+        fullPageDataBuffer={fullPageDataBuffer}
+        serverConsumerManifest={serverConsumerManifest}
+        segmentPath={segmentPath}
+      />,
+      clientModules
+    )
+    const segmentBuffer = await streamToBuffer(segmentStream)
+    // Add the buffer to the result map.
+    if (segmentPathStr === '') {
+      segmentBufferMap.set('/', segmentBuffer)
+    } else {
+      segmentBufferMap.set(segmentPathStr, segmentBuffer)
+    }
+  } catch {
+    // If there are any errors, then we skip the segment. The effect is that
+    // a prefetch for this segment will 404.
+  }
+}
+
+async function PickSegment({
+  fullPageDataBuffer,
+  serverConsumerManifest,
+  segmentPath,
+}: {
+  fullPageDataBuffer: Buffer
+  serverConsumerManifest: any
+  segmentPath: Array<[string, Segment]>
+}): Promise<any> {
+  // We're currently rendering a Flight response for a segment prefetch.
+  // Decode the Flight stream for the whole page, then pick out the data for the
+  // segment at the given path. This ends up happening once per segment. Not
+  // ideal, but we do it this way so that that we can transfer the side effects
+  // from the original Flight stream (e.g. Float preloads) onto the Flight
+  // stream for each segment's prefetch.
+  //
+  // This does mean that a prefetch for an individual segment will include the
+  // resources for the entire page it belongs to, but this is a reasonable
+  // trade-off for now. The main downside is a bit of extra bandwidth.
+  const replayConsoleLogs = true
+  const rscPayload: InitialRSCPayload = await createFromReadableStream(
```

Comment:
@gnoff What happens in the case where there's a dynamic data access outside of a Suspense boundary? Does this error or does the promise just never resolve?

Reply:
On second thought, my question doesn’t make sense, so I’ll rephrase to what I actually wanted to know: is there any situation where this promise would hang indefinitely and never resolve?

Reply:
Nvm I figured it out, I forgot that the dynamic holes are encoded in the result stream as hanging promises